Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,178 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Extensibility.Implementation.Tracing;
+    /// <summary>
+    /// A store for instrumentation App Ids. This makes sure we don't query the public endpoint to find an app Id for the same Ikey more than once.
+    /// </summary>
+    internal class CorrelationIdLookupHelper
+    {
+        /// <summary>
+        /// Max number of app ids to cache.
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        // For now we have decided to go with not waiting to retrieve the app Id, instead we just cache it on retrieval.
+        // This means the initial few attempts to get correlation id might fail and the inital telemetry sent might be missing such data.
+        // However, once it is in the cache - subsequent telemetry should contain this data. 
+        private int getAppIdTimeout = 0; // milliseconds
+
+        private const string CORRELATION_ID_FORMAT = "aid-v1:{0}";
+
+        private const string APPID_QUERY_API_RELATIVE_URI_FORMAT = "api/profiles/{0}/appId";
+
+        private Uri endpointAddress;
+
+        private ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
+
+        private Func<string, Task<string>> provideAppId;
+
+        /// <summary>
+        /// This constructor is mostly used by the test classes to provide an override for fetching appId logic
+        /// </summary>
+        /// <param name="appIdProviderMethod">The delegate to be called to fetch the appId</param>
+        /// <param name="expectedResponseTime">Wait time for the provided method to return the app Id.</param>
+        public CorrelationIdLookupHelper(Func<string, Task<string>> appIdProviderMethod, int expectedResponseTime = 20)
```

Comment:
In fact - it just so happens that my mock tasks are created - completed. So I don't even need a timeout. I'm getting rid of the timeout input parameter on the constructor.